genfs.c

```diff
@@ -0,0 +1,324 @@
+            mark_immutable = true;
+        else if (!strcmp("immutable=no", argv[3]))
+            mark_immutable = false;
+        else
+            goto misuse;
+    } else if (argc != 2) {
+misuse:
+        errx(1, "Usage: genfs FILESYSTEM [UNAME immutable=[yes|no]]");
+    }
+    const char *const filesystem = argv[1], *const uname = argv[2];
+    return real_main(filesystem, uname);
+}
+
+struct qubes_genfs_data {
+    ext2_filsys fs;
+    const char *const uname;
+};
+
+static int recursive_relabel(ext2_ino_t dir __attribute__((unused)),
+                             int entry __attribute__((unused)),
+                             struct ext2_dir_entry *dirent,
+                             int offset __attribute__((unused)),
+                             int blocksize __attribute__((unused)),
+                             char *buf __attribute__((unused)),
+                             void *priv_data);
+
+static int root_iterate_callback(ext2_ino_t dir __attribute__((unused)),
+                                 int entry __attribute__((unused)),
+                                 struct ext2_dir_entry *dirent,
+                                 int offset __attribute__((unused)),
+                                 int blocksize __attribute__((unused)),
+                                 char *buf __attribute__((unused)),
+                                 void *priv_data) {
+    struct qubes_genfs_data *data = priv_data;
+    int const name_len = ext2fs_dirent_name_len(dirent);
+    errcode_t err;
+    process_dirent(data->fs, "", dirent->inode, name_len, dirent->name,
+                   "system_u:object_r:modules_object_t:s0", mark_immutable);
+    if (!strncmp(dirent->name, "firmware", (size_t)name_len)) {
+        return 0;
+        struct qubes_genfs_data relabel_data = {
+            .fs = data->fs,
+            .uname = NULL,
+        };
+        if ((err = ext2fs_dir_iterate2(data->fs, dirent->inode, 0, NULL, recursive_relabel, &relabel_data)))
+            genfs_err(err, "during recursive relabel");
+    } else if (!strncmp(dirent->name, data->uname, (size_t)name_len)) {
+        if ((err = ext2fs_dir_iterate2(data->fs, dirent->inode, 0, NULL, dir_iterate_callback, data)))
+            genfs_err(err, "processing %s", data->uname);
+    }
+    return 0;
+}
+
+static int dir_iterate_callback(ext2_ino_t dir __attribute__((unused)),
+                                int entry __attribute__((unused)),
+                                struct ext2_dir_entry *dirent,
+                                int offset __attribute__((unused)),
+                                int blocksize __attribute__((unused)),
+                                char *buf __attribute__((unused)),
+                                void *priv_data) {
+    struct qubes_genfs_data *data = priv_data;
+    int name_len = ext2fs_dirent_name_len(dirent);
+    assert(name_len >= 0 && name_len < EXT2_NAME_LEN);
+    _Static_assert(EXT2_NAME_LEN >= 255, "libext2fs bug");
```

Comment:
Do you actually need this property anywhere? Otherwise please drop - using a proper library (header) is about _avoiding_ open-coded constants, not duplicating them.